beacon-chain/sync/data_columns.go

```diff
@@ -0,0 +1,444 @@
+	dataColumns map[uint64]bool,
+	block interfaces.ReadOnlySignedBeaconBlock,
+	blkRoot [32]byte,
+	peers []core.PeerID,
+	clock *startup.Clock,
+	p2p p2p.P2P,
+	ctxMap ContextByteVersions,
+	newColumnsVerifier verification.NewDataColumnsVerifier,
+) ([]blocks.RODataColumn, error) {
+	if len(dataColumns) == 0 {
+		return nil, nil
+	}
+
+	// Assemble the peers who can provide the needed data columns.
+	dataColumnsByAdmissiblePeer, _, _, err := AdmissiblePeersForDataColumns(peers, dataColumns, p2p)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get admissible peers for data columns")
+	}
+
+	sidecars := make([]blocks.RODataColumn, 0, len(dataColumns))
+	remainingColumns := make(map[uint64]bool, len(dataColumns))
+	for col := range dataColumns {
+		remainingColumns[col] = true
+	}
+
+	for len(dataColumnsByAdmissiblePeer) > 0 {
+		// Try to select peers excluding bad peers
+		peersToFetchFrom, err := SelectPeersToFetchDataColumnsFrom(remainingColumns, dataColumnsByAdmissiblePeer)
+		if err != nil {
+			// Return an error if some columns are unavailable from the filtered set
+			// of peers. Filtering out bad peers can make columns unavailable, and
+			// when that happens, the caller needs to know about it.
+			return nil, errors.Wrap(err, "couldn't select peers to fetch data columns from")
+		}
+
+		// Request the data columns from each peer
+		successfulColumns := make(map[uint64]bool)
+		for peer, dataColumns := range peersToFetchFrom {
+			request, err := RequestsForDataColumnsByRoot(blkRoot, dataColumns)
+			if err != nil {
+				log.WithError(err).Debug("Failed to build request for data columns")
+				continue
+			}
+
+			peerSidecars, err := SendDataColumnSidecarsByRootRequest(ctx, clock, p2p, peer, ctxMap, &request)
+			if err != nil {
+				// Remove this peer since it failed to respond correctly
+				delete(dataColumnsByAdmissiblePeer, peer)
+				log.WithFields(logrus.Fields{
+					"peer":      peer.String(),
+					"blockRoot": fmt.Sprintf("%#x", blkRoot),
+					"error":     err.Error(),
+				}).Debug("Failed to request data columns from peer")
+				continue
+			}
+
+			// Mark columns as successful
+			for _, sidecar := range peerSidecars {
+				colIndex := sidecar.ColumnIndex
+				successfulColumns[colIndex] = true
+			}
+
+			for _, colIndex := range dataColumns {
+				if !successfulColumns[colIndex] {
+					// Remove this peer if any requested column wasn't successful
+					delete(dataColumnsByAdmissiblePeer, peer)
+					log.WithFields(logrus.Fields{
+						"peer":          peer.String(),
+						"missingColumn": colIndex,
+					}).Debug("Peer failed to return requested data column")
+					break
+				}
+			}
+
+			sidecars = append(sidecars, peerSidecars...)
+		}
+
+		// Update remaining columns for the next retry
+		for col := range successfulColumns {
+			delete(remainingColumns, col)
+		}
+
+		if len(remainingColumns) > 0 {
+			// Some columns are still missing, retry with the remaining peers.
+			continue
```

Comment:
This case is not tested (correlated with the fact that a peer did not give at least one requested column). As said before, I think it is important to test it.

(So you probably need, in the test configuration, to tell one peer what columns it should respond.)

Reply:
[a93bdf7](https://github.com/prysmaticlabs/prysm/pull/14977/commits/a93bdf7ac2a69f8e2dadf8ca2e8f7c55ce501a48)